File: Data Structures & Algorithms/longest-repeating-substring-with-replacement/submission-0.py
class Solution:
    def characterReplacement(self, s: str, k: int) -> int:
        mp = {}

        maxi = 0
        l = 0
        r = 0
        n = len(s)
        maxFreq = 0

        while r < n:
            mp[s[r]] = 1 + mp.get(s[r], 0)
            maxFreq = max(maxFreq, mp[s[r]])

            while (r - l + 1) - maxFreq > k:
                mp[s[l]] -= 1
                l += 1
            
            maxi = max(maxi, r - l + 1)
            r += 1
        
        return maxi

File: Data Structures & Algorithms/house-robber-ii/submission-0.cpp
class Solution {
public:
    int robLinear(vector<int>& nums, int st, int end) {
        int n = nums.size();
        int prev2 = 0, prev1 = 0, curr;
        for(int i = st; i <= end; i++) {
            curr = max(prev1, prev2 + nums[i]);
            prev2 = prev1;
            prev1 = curr;
        }
        return prev1;
    }
    int rob(vector<int>& nums) {
        int n = nums.size();
        if(n == 0) return 0;
        if(n == 1) return nums[0];
        if(n == 2) return max(nums[0], nums[1]);
        
        int excludeFirst = robLinear(nums, 1, n - 1);
        int excludeLast = robLinear(nums, 0, n - 2);

        return max(excludeFirst, excludeLast);
    }
};
